compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -0,0 +1,93 @@
+//! Deduces supplementary parameter attributes from MIR.
+//!
+//! Deduced parameter attributes are those that can only be soundly determined by examining the
+//! body of the function instead of just the signature. These can be useful for optimization
+//! purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+//! dependent crates can use them.
+
+use rustc_hir::def_id::DefId;
+use rustc_index::bit_set::BitSet;
+use rustc_middle::mir::visit::{PlaceContext, Visitor};
+use rustc_middle::mir::{Body, Local, Location};
+use rustc_middle::ty::{DeducedParamAttrs, ParamEnv, TyCtxt};
+use rustc_span::DUMMY_SP;
+
+// A visitor that determines which arguments have been mutated. We can't use the mutability field on
+// LocalDecl for this because it has no meaning post-optimization.
+struct DeduceReadOnly {
+    // Each bit is indexed by argument number, starting at zero (so 0 corresponds to local decl 1).
+    // The bit is true if the argument may have been mutated or false if we know it hasn't been up
+    // to the point we're at.
+    mutable_args: BitSet<usize>,
+}
+
+impl DeduceReadOnly {
+    // Returns a new DeduceReadOnly instance.
+    fn new(arg_count: usize) -> Self {
+        Self { mutable_args: BitSet::new_empty(arg_count) }
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for DeduceReadOnly {
+    fn visit_local(&mut self, local: Local, context: PlaceContext, _: Location) {
+        // We're only interested in arguments.
+        if local.index() == 0 || local.index() > self.mutable_args.domain_size() {
+            return;
+        }
+
+        match context {
+            PlaceContext::NonMutatingUse(..) | PlaceContext::NonUse(..) => {
+                // Not mutating, so it's fine.
+            }
+            PlaceContext::MutatingUse(..) => {
+                // This is a mutation, so mark it as such.
+                self.mutable_args.insert(local.index() - 1);
+            }
+        }
+    }
+}
+
+/// Returns the deduced parameter attributes for a function.
+///
+/// Deduced parameter attributes are those that can only be soundly determined by examining the
+/// body of the function instead of just the signature. These can be useful for optimization
+/// purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+/// dependent crates can use them.
+pub fn deduced_param_attrs<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> &'tcx [DeducedParamAttrs] {
+    // For now, don't deduce any parameter attributes on intrinsics.
+    //
+    // FIXME(pcwalton): Deduce some?
+    if !tcx.is_closure(def_id) && tcx.is_intrinsic(def_id) {
+        return &[];
+    }
+
+    // Don't deduce any parameter attributes on default methods.
```

Comment:
Why is this a problem?

Reply:
Ah, so the FIXME is for default methods, but the main comment for trait methods without default bodies, as fetching their body panics.

Reply:
Gotcha, I updated the comment, if I understand you properly.